### PR TITLE
Support retry for WRITE_THROUGH and WRITE_BEHIND 

### PR DESCRIPTION
**Pull request by zzhlhc**
Related #5224 , first commits are only for discussing feasibility,and i will complete it if it's suitable.

- #### Use new `MapWriterAsync` constructor to enable retry (original constructor doesn't enable retry):
```
        new MapWriterAsync(retryAttempts:3, retryInterval:100){
            @Override
            public CompletionStage<Void> write(Map values){
                for (Map.Entry entry : values.entrySet()) {
                    db.insert(entry);
                    //write success so this entry won't participate in future retries even if exception thrown in "values.entrySet()" loop
                    writeSuccess(entry);
                }
            }

            //another override
            @Override
            public CompletionStage<Void> write(Map values){
                db.insertAll(valuse);
                writeSuccess(values); 
            }
            
        }
```
I think only user self knows what is write/delete success, so we can provide writeSuccess/deleteSuccess methods, users call them to avoid invalid retries.

- #### Some changes in the code
1. Add `abstract class RetryableWriter` to provide information needed for retry.
2. Change `MapWriter/MapWriterAsync` to abstract class so they can extends `RetryableWriter` and become retryable, `RetryableWriter` can also be extended by other `Writer`s if needed in the future.
3. Add retry logic in `MapWriteBehindTask::flushTasks` (`processTask` not yet).

First commits are only for discussing feasibility,and i will complete it if it's suitable.

### CODE REVIEW

**Review by mrniko — CHANGES_REQUESTED** [review]
Thank you for the proposed changes. writerRetryAttempts and writerRetryAttempts parameters should belong to MapOptions object.

**Comment by zzhlhc** [comment]
Okay! I'll finish this in two days.

**Comment by mrniko** [comment]
The better way is implement MapWriter and MapWriterAsync delegators with retry logic inside.

**Comment by zzhlhc** [comment]
Pushed with some confusion:
```
public class RetryableMapWriterAsync<K, V> implements MapWriterAsync<K, V> {
    private final MapWriterAsync<K, V> mapWriterAsync;
}
```
In the above code, i can't provide `writeSuccess` in `RetryableMapWriterAsync`, because users only instantiate `MapWriterAsync` so they won't trigger `RetryableMapWriterAsync::writeSuccess`.

Any suggestions or should i give up `writeSuccess` method?

File: redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        while (leftAddAttempts > 0) {
+            try {
+                //remove successful part
+                if (!noRetriesForWrite.isEmpty()) {
+                    noRetriesForWrite.forEach(addedMap::remove);
+                    noRetriesForWrite.clear();
+                }
+
+                //do write
+                return mapWriterAsync.write(addedMap);
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java:56]: This is incorrect since error reported by CompletionStage and not write() method

File: redisson/src/main/java/org/redisson/api/map/MapWriterAsync.java [hunk 2]
```diff
@@ -21,7 +21,7 @@
 
 /**
  * Asynchronous Map writer used for write-through operations.
- * 
+ *
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/map/MapWriterAsync.java:24]: remove unnecessary changes.

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/api/map/MapWriterAsync.java (hunk 2 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/api/map/MapWriterAsync.java]: done

File: redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java (hunk 1 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java]: My mistake, corrected.

File: redisson/src/main/java/org/redisson/MapWriteBehindTask.java [hunk 3]
```diff
@@ -100,7 +100,7 @@ private void flushTasks(Map<Object, Object> addedMap, List<Object> deletedKeys)
                 if (options.getWriter() != null) {
                     options.getWriter().write(addedMap);
                 } else {
-                    options.getWriterAsync().write(addedMap).toCompletableFuture().join();
+                    options.getWriterAsync().write(addedMap);
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/MapWriteBehindTask.java:103]: join should be retained since write method is executed in a separate thread.

**Review by mrniko — CHANGES_REQUESTED** [review]
I think retry logic should be applied to MapWriter as well.

File: redisson/src/main/java/org/redisson/api/MapOptions.java [hunk 4]
```diff
@@ -159,7 +165,25 @@ public MapOptions<K, V> writeMode(WriteMode writeMode) {
     public WriteMode getWriteMode() {
         return writeMode;
     }
-    
+
+    public int getWriterRetryAttempts() {
+        return writerRetryAttempts;
+    }
+
+    public MapOptions<K, V> writerRetryAttempts(int writerRetryAttempts) {
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/MapOptions.java:173]: Please add comments

File: redisson/src/main/java/org/redisson/api/MapOptions.java [hunk 5]
```diff
@@ -159,7 +165,25 @@ public MapOptions<K, V> writeMode(WriteMode writeMode) {
...
+        this.writerRetryAttempts = writerRetryAttempts;
+        return this;
+    }
+
+    public long getWriterRetryInterval() {
+        return writerRetryInterval;
+    }
+
+    public MapOptions<K, V> writerRetryInterval(long writerRetryInterval, TimeUnit timeUnit) {
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/MapOptions.java:182]: Please add comments

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/MapWriteBehindTask.java (hunk 3 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/MapWriteBehindTask.java:103]: Sorry, i don't quite understand this comment.

The line at 103 where you commented will call `RetryableMapWriterAsync::write`, which is executed in the same thread as `flushTasks`. Then, step in `RetryableMapWriterAsync:56` : `mapWriterAsync.write(addedMap).toCompletableFuture().join(); ` it has joined to wait for a separate thread.

So, do you mean I should let `RetryableMapWriterAsync:write` look like `return CompletableFuture.runAsync(() -> {//retry logic})`? Because only this way can help me understand this requested change.

File: redisson/src/main/java/org/redisson/api/MapOptions.java (hunk 5 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/api/MapOptions.java:199]: Done.

File: redisson/src/main/java/org/redisson/api/MapOptions.java (hunk 4 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/api/MapOptions.java:180]: Done.

File: redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    @Override
+    public CompletionStage<Void> write(Map<K, V> addedMap) {
+        return CompletableFuture.runAsync(() -> {
+            //execute at least once
+            int leftAddAttempts = Math.max(1, options.getWriterRetryAttempts());
+            while (leftAddAttempts > 0) {
+                try {
+                    //do write
+                    mapWriterAsync.write(addedMap).toCompletableFuture().join();
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java:47]: in async way you need to schedule next attempt and not use sleep method. Someday write() method will be invoked in async way only, so synchronization should be avoided.

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java (hunk 6 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java]: Done.

Don't know if my changes are reasonable so only `RetryableMapWriterAsync:write` and `MapWriteBehindTask:flushTasks` are modified, please check.

**Comment by zzhlhc** [comment]
@mrniko hi, any review?

File: redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public CompletionStage<Void> write(Map<K, V> addedMap) {
+        return retryWrite(Math.max(1, options.getWriterRetryAttempts()), new LinkedHashMap<>(addedMap));
+    }
+
+    private CompletableFuture<Void> retryWrite(int leftAttempts, Map<K, V> addedMap) {
+        return CompletableFuture.runAsync(() -> {
+            try {
+                //do write
+                mapWriterAsync.write(addedMap).toCompletableFuture().join();
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java:55]: Call is still sync. Can you do it without CompletableFuture.runAsync ? You can look at org.redisson.RedissonLock#lockAsync(long, java.util.concurrent.TimeUnit, org.redisson.RedissonLockEntry, java.util.concurrent.CompletableFuture<java.lang.Void>, long) method as example.

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java (hunk 7 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/api/map/RetryableMapWriterAsync.java]: Done!

File: redisson/src/main/java/org/redisson/api/MapOptions.java (hunk 5 above)
Comment (mrniko) [reply redisson/src/main/java/org/redisson/api/MapOptions.java:199]: Please change writerRetryInterval type to Duration

File: redisson/src/main/java/org/redisson/api/MapOptions.java [hunk 8]
```diff
@@ -89,7 +96,7 @@ public static <K, V> MapOptions<K, V> defaults() {
      * @return MapOptions instance
      */
     public MapOptions<K, V> writer(MapWriter<K, V> writer) {
-        this.writer = writer;
+        this.writer = new RetryableMapWriter<>(this, writer);
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/MapOptions.java:99]: don't apply RetryableMapWriter if writerRetryAttempts == 0

File: redisson/src/main/java/org/redisson/api/MapOptions.java [hunk 9]
```diff
@@ -103,7 +110,7 @@ public MapWriter<K, V> getWriter() {
      * @return MapOptions instance
      */
     public MapOptions<K, V> writerAsync(MapWriterAsync<K, V> writer) {
-        this.writerAsync = writer;
+        this.writerAsync = new RetryableMapWriterAsync<>(this, writer);
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/MapOptions.java:113]: don't apply RetryableMapWriter if writerRetryAttempts == 0

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/api/MapOptions.java (hunk 8 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/api/MapOptions.java:99]: One question: when create MapOptions,  `.writerRetryAttempts()` maybe called after`.writerAsync()`, like this:

```
           MapOptions.<>defaults()
                .writerAsync(new MapWriterAsync<String, String>() {
                 ...
                })
                .writerRetryAttempts(expectedRetryAttempts);
```

In this case, `RetryableMapWriter` won't be applied even if users intend it to be.

File: redisson/src/main/java/org/redisson/api/MapOptions.java [hunk 10]
```diff
@@ -89,7 +96,11 @@ public static <K, V> MapOptions<K, V> defaults() {
      * @return MapOptions instance
      */
     public MapOptions<K, V> writer(MapWriter<K, V> writer) {
-        this.writer = writer;
+        if (writerRetryAttempts ==0) {
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/api/MapOptions.java:99]: I thought you will remove that check since you were right - writerRetryAttempts can be set later

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/api/MapOptions.java (hunk 10 above)
Comment (zzhlhc) [reply redisson/src/main/java/org/redisson/api/MapOptions.java]: I didn't remove it because I didn't get your confirmation. Now I got it and changed it back.

**Comment by mrniko** [comment]
Thanks for contribution!
